### PR TITLE
[1.21] Add NamespacedDirectoryLister

### PR DESCRIPTION
**Pull request by XFactHD**
This PR adds the `NamespacedDirectoryLister`, a namespace-aware version of vanilla's `DirectoryLister` sprite source.
The motivation behind that is the regular occurence of mods adding textures to a generic, high-traffic directory (common examples are `textures/model(s)/` and `textures/entity/`) and then using a `DirectoryLister` (which is not namespace-aware and will collect files in the specified directory and sub-directories thereof from all namespaces) to stitch their textures to the block atlas. When this is done for directories like the two examples (especially the latter) which often contain entity-style texture maps which that a very large footprint, then the block atlas quickly gets blown up to absurd sizes, sometimes even blowing past texture size limits of certain GPUs (8k<sup>2</sup> for older GPUs, 16k<sup>2</sup> for the large majority of GPUs and 32k<sup>2</sup> for only the latest Nvidia GPUs) when a lot of mods or particularly big mods (in terms of textures in the affected directories) are involved.
It's worth noting that patching the vanilla `DirectoryLister` sprite source to be namespace-aware by default (i.e. without adding an optional namespace parameter) is not an option for two reasons:
1. A sprite source does not know which resource pack it came from and funneling that data through is painful ([ask me how I know](https://www.curseforge.com/minecraft/mc-mods/atlasviewer))
2. Even if it were aware of its origin, doing so would break any and all assumptions about `textures/block/` and `textures/item/` working out of the box on the block atlas (the same of course applies to basically all other atlases as well) and would therefore break all mods and resourcepacks that place textures in these directories outside of the vanilla namespace.

---

I also took the liberty to make the `RegisterSpriteSourceTypesEvent` less of a pain to use as there is no reason to not construct the `SpriteSourceType` outside the event since it's just a paper-thin wrapper around the codec. The current implementation makes storage and use of the type unnecessarily annoying.

### CODE REVIEW

File: src/main/java/net/neoforged/neoforge/client/event/RegisterSpriteSourceTypesEvent.java [hunk 1]
```diff
@@ -39,13 +39,26 @@ public RegisterSpriteSourceTypesEvent(BiMap<ResourceLocation, SpriteSourceType>
...
     }
+
+    /**
+     * Registers the given {@link SpriteSourceType} under the given id.
+     *
+     * @param id         The id to register the {@link SpriteSourceType} under
+     * @param sourceType The {@link SpriteSourceType} to register
+     */
+    public void register(ResourceLocation id, SpriteSourceType sourceType) {
+        this.types.put(id, sourceType);
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/event/RegisterSpriteSourceTypesEvent.java:65]: Shouldn't this do the check against duplicate sprite source registration, as the now-deprecated method does?

File: src/main/java/net/neoforged/neoforge/client/textures/NamespacedDirectoryLister.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+ */
+public record NamespacedDirectoryLister(String namespace, String sourcePath, String idPrefix) implements SpriteSource {
+
+    private static final MapCodec<NamespacedDirectoryLister> CODEC = RecordCodecBuilder.mapCodec(inst -> inst.group(
+            Codec.STRING.fieldOf("namespace").forGetter(lister -> lister.namespace),
+            Codec.STRING.fieldOf("source").forGetter(lister -> lister.sourcePath),
+            Codec.STRING.fieldOf("prefix").forGetter(lister -> lister.idPrefix)).apply(inst, NamespacedDirectoryLister::new));
+    public static final ResourceLocation ID = ResourceLocation.fromNamespaceAndPath(NeoForgeVersion.MOD_ID, "namespaced_directory");
+    public static final SpriteSourceType TYPE = new SpriteSourceType(CODEC);
+    @Override
```
Comment (sciwhiz12) [inline src/main/java/net/neoforged/neoforge/client/textures/NamespacedDirectoryLister.java:32]: Please add a blank line between the field and the method.

**Review by sciwhiz12 — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/client/event/RegisterSpriteSourceTypesEvent.java (hunk 1 above)
Comment (XFactHD) [reply src/main/java/net/neoforged/neoforge/client/event/RegisterSpriteSourceTypesEvent.java:65]: Yes, it definitely should

File: src/main/java/net/neoforged/neoforge/client/textures/NamespacedDirectoryLister.java (hunk 2 above)
Comment (XFactHD) [reply src/main/java/net/neoforged/neoforge/client/textures/NamespacedDirectoryLister.java:32]: I can't, just as I can't remove the empty line between the record header and the first field, blame Spotless.

File: tests/src/main/java/net/neoforged/neoforge/debug/client/ClientTests.java [hunk 3]
```diff
@@ -81,6 +84,27 @@ static void keyMappingTest(final DynamicTest test) {
         });
     }
 
+    @TestHolder(description = "Tests that the NamespacedDirectoryLister only collects resources from the specified namespace")
```
Comment (Matyrobbrt) [inline tests/src/main/java/net/neoforged/neoforge/debug/client/ClientTests.java]: This test should be marked as enabled by default for the fail/pass to actually work.

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

**Review by sciwhiz12 — APPROVED** [review]
Looks fine to me, and the test passes. Waiting on @Matyrobbrt to review the changes to the test framework.

**Review by Matyrobbrt — APPROVED** [review]

**Pull request merged**